src/blocked.c

```diff
@@ -863,10 +466,196 @@ static void signalKeyAsReadyLogic(redisDb *db, robj *key, int type, int deleted)
+ *    and instead clear the dictionary in one-shot. this is why the remove_key argument is provided
+ *    to support this logic in unblockClientWaitingData
+ */
+static void releaseBlockedEntry(client *c, dictEntry *de, int remove_key) {
+    list *l;
+    listNode *pos;
+    void *key;
+    dictEntry *unblock_on_nokey_entry;
+
+    key = dictGetKey(de);
+    pos = dictGetVal(de);
+    /* Remove this client from the list of clients waiting for this key. */
+    l = dictFetchValue(c->db->blocking_keys, key);
+    serverAssertWithInfo(c,key,l != NULL);
+    listUnlinkNode(l,pos);
+    /* If the list is empty we need to remove it to avoid wasting memory
+     * We will also remove the key (if exists) from the blocking_keys_unblock_on_nokey dict.
+     * However, in case the list is not empty, we will have to still perform reference accounting
+     * on the blocking_keys_unblock_on_nokey and delete the entry in case of zero reference.
+     * Why? because it is possible that some more clients are blocked on the same key but without
+     * require to be triggered on key deletion, we do not want these to be later triggered by the
+     * signalDeletedKeyAsReady. */
+    if (listLength(l) == 0) {
+        dictDelete(c->db->blocking_keys, key);
+        dictDelete(c->db->blocking_keys_unblock_on_nokey,key);
+    } else if (c->bstate.unblock_on_nokey) {
+        unblock_on_nokey_entry = dictFind(c->db->blocking_keys_unblock_on_nokey,key);
+        /* it is not possible to have a client blocked on nokey with no matching entry */
+        serverAssertWithInfo(c,key,unblock_on_nokey_entry != NULL);
+        if (!dictIncrUnsignedIntegerVal(unblock_on_nokey_entry, -1)) {
+            /* in case the count is zero, we can delete the entry */
+             dictDelete(c->db->blocking_keys_unblock_on_nokey,key);
+        }
+    }
+    if (remove_key)
+        dictDelete(c->bstate.keys, key);
+}
+
 void signalKeyAsReady(redisDb *db, robj *key, int type) {
     signalKeyAsReadyLogic(db, key, type, 0);
 }
 
 void signalDeletedKeyAsReady(redisDb *db, robj *key, int type) {
     signalKeyAsReadyLogic(db, key, type, 1);
 }
+
+/* Helper function for handleClientsBlockedOnKeys(). This function is called
+ * whenever a key is ready. we iterate over all the clients blocked on this key
+ * and try to re-execute the command (in case the key is still available). */
+static void handleClientsBlockedOnKey(readyList *rl) {
+
+    /* We serve clients in the same order they blocked for
+     * this key, from the first blocked to the last. */
+    dictEntry *de = dictFind(rl->db->blocking_keys,rl->key);
+
+    if (de) {
+        list *clients = dictGetVal(de);
+        listNode *ln;
+        listIter li;
+        listRewind(clients,&li);
+
+        while((ln = listNext(&li))) {
+            client *receiver = listNodeValue(ln);
+            robj *o = lookupKeyReadWithFlags(rl->db, rl->key, LOOKUP_NONOTIFY | LOOKUP_NOSTATS | LOOKUP_NOTOUCH | LOOKUP_NOEXPIRE);
```

Comment:
use `dictFind` directly here?

Reply:
I think lookupKeyReadWithFlags also cover some checks (i.e expiry checking) but maybe it would have been a good idea to present a new type of enu(LOOKUP_TRANSPARENT) which we will always maintain to do nothing with the key
like 
#define LOOKUP_TRANSPARENT (LOOKUP_NONOTIFY | LOOKUP_NOSTATS | LOOKUP_NOTOUCH | LOOKUP_NOEXPIRE)

Reply:
i don't think we should use dictFind. we do want to know if the key is logically there (e.g. we do want an expiration test, just not to delete the key if it is expired).
it's essentially a lookup key with 0 side effects.
maybe want to add a LOOKUP_NO_SIDEEFFECTS and use it here, this way other flags added in the future are less likely to be overlooked.

Reply:
done